### PR TITLE
Fix empty bytes

### PR DESCRIPTION
**Pull request by chaokunyang**
## Why are these changes needed?
This is for fix `Cannot have empty byte string (b"") argument for remote calls` in issue #6944

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

**Comment by simon-mo** [comment]
Can you add a test in python/ray/tests/test_serialization.py to make sure this doesn’t break in the future? Thanks!

File: src/ray/common/buffer.h [hunk 1]
```diff
@@ -67,9 +67,10 @@ class LocalMemoryBuffer : public Buffer {
 
   /// Construct a LocalMemoryBuffer of all zeros of the given size.
   LocalMemoryBuffer(size_t size) : has_data_copy_(true) {
-    buffer_.resize(size, 0);
+    // ensure data_ won't be null_ptr when size == 0
+    buffer_.resize(std::max<size_t>(1, size), 0);
```
Comment (edoakes) [inline src/ray/common/buffer.h]: This seems like a bit of a hacky workaround to me. Is the issue that the `std::vector` might have a `nullptr` underlying if size is 0? Maybe we should just enforce that `data_` is `nullptr` if `size_ == 0` instead and handle that elsewhere?

Comment (chaokunyang) [reply src/ray/common/buffer.h]: How about handle in `write_serialized_object`:
```
    if isinstance(serialized_object, RawSerializedObject):
        if buf.get().Size() > 0:
            buffer = Buffer.make(buf)
            # `pyarrow.py_buffer` will crash when underlying buffer is a null_ptr
            stream = pyarrow.FixedSizeBufferWriter(pyarrow.py_buffer(buffer))
            stream.set_memcopy_threads(MEMCOPY_THREADS)
            stream.write(pyarrow.py_buffer(serialized_object.value))
```

File: src/ray/core_worker/core_worker.cc [hunk 2]
```diff
@@ -887,17 +887,16 @@ Status CoreWorker::AllocateReturnObjects(
...
-          static_cast<int64_t>(data_sizes[i]) <
-              RayConfig::instance().max_direct_call_object_size() &&
-          contained_object_ids[i].empty()) {
-        data_buffer = std::make_shared<LocalMemoryBuffer>(data_sizes[i]);
-      } else {
-        RAY_RETURN_NOT_OK(Create(metadatas[i], data_sizes[i], contained_object_ids[i],
-                                 object_ids[i], &data_buffer));
-        object_already_exists = !data_buffer;
-      }
+    // when data_sizes[i] == 0, we should Allocate an empty buffer
```
Comment (jovany-wang) [inline src/ray/core_worker/core_worker.cc]: No need to handle it here.

I think it's ok to fix it at `_raylet.pyx` when `desrialize_args` right?

Comment (chaokunyang) [reply src/ray/core_worker/core_worker.cc]: After fix it in `_raylet.pyx`, I came across another problem:
```
ray.exceptions.RayWorkerError: The worker died unexpectedly while executing this task.
(pid=36089) *** Aborted at 1580884257 (unix time) try "date -d @1580884257" if you are using GNU date ***
(pid=36089) PC: @                0x0 (unknown)
(pid=36089) *** SIGSEGV (@0x0) received by PID 36089 (TID 0x7fffa970c380) stack trace: ***
(pid=36089)     @     0x7fff71365f5a _sigtramp
(pid=36089)     @        0x10ea02db8 (unknown)
(pid=36089)     @        0x104040abe std::__1::__function::__func<>::operator()()
(pid=36089)     @        0x10406ddb2 ray::CoreWorkerDirectTaskReceiver::HandlePushTask()::$_2::operator()()
(pid=36089)     @        0x10406cfcc ray::CoreWorkerDirectTaskReceiver::HandlePushTask()
(pid=36089)     @        0x10404daee boost::asio::detail::completion_handler<>::do_complete()
(pid=36089)     @        0x10403953f boost::asio::detail::scheduler::do_run_one()
(pid=36089)     @        0x104039052 boost::asio::detail::scheduler::run()
(pid=36089)     @        0x104017b42 ray::CoreWorker::StartExecutingTasks()
(pid=36089)     @        0x103fac1ab __pyx_pw_3ray_7_raylet_10CoreWorker_3run_task_loop()
(pid=36089)     @        0x102c55ae0 _PyCFunction_FastCallDict
(pid=36089)     @        0x102cd7927 call_function
(pid=36089)     @        0x102cd40d2 _PyEval_EvalFrameDefault
(pid=36089)     @        0x102cd8b9d fast_function
(pid=36089)     @        0x102cd7901 call_function
(pid=36089)     @        0x102cd40d2 _PyEval_EvalFrameDefault
(pid=36089)     @        0x102cd838f _PyEval_EvalCodeWithName
(pid=36089)     @        0x102ccd534 PyEval_EvalCode
(pid=36089)     @        0x102d03511 PyRun_FileExFlags
(pid=36089)     @        0x102d02cc3 PyRun_SimpleFileExFlags
(pid=36089)     @        0x102d1bc49 Py_Main
(pid=36089)     @        0x102bf7df8 main
(pid=36089)     @     0x7fff71057015 start
(pid=36089)     @                0xc (unknown)
```

Comment (chaokunyang) [reply src/ray/core_worker/core_worker.cc]: The code scattered in other places not only in `deserialize_args`.

If `data_sizes[i] == 0`, then  buffer of `RayObject` for `return_objects[i]` will be an null_ptr. We need extra check in client code.

> Can you add a test in python/ray/tests/test_serialization.py to make sure this doesn’t break in the future? Thanks!

add two bytes values in `test_simple_serialization`

Comment (chaokunyang) [reply src/ray/core_worker/core_worker.cc]: fixed

**Comment by chaokunyang** [comment]
> Can you add a test in python/ray/tests/test_serialization.py to make sure this doesn’t break in the future? Thanks!

add two bytes values in `test_simple_serialization`

**Review by jovany-wang — APPROVED** [review]
LGTM

**Comment by jovany-wang** [comment]
CI looks unrelated.

**Pull request merged**